ssl/ssl_local.h

```diff
@@ -1799,7 +1808,25 @@ struct ssl_st {
      */
     const struct sigalg_lookup_st **shared_sigalgs;
     size_t shared_sigalgslen;
-};
+} SSL_CONNECTION;
+
+# define SSL_CONNECTION_FROM_SSL_ONLY(sslcon, ssl, ret_stmt) \
```

Comment:
Wondering if a "function" like form might be more readable:
```c
# define SSL_CONNECTION_FROM_SSL_ONLY(ssl) \
    (ssl->type == SSL_TYPE_SSL_CONNECTION  \
     ? (SSL_CONNECTION *)ssl               \
     : NULL)
```

Reply:
It's possible alternative however it means there will have to be an if statement at the beginning of every SSL function. But yeah, that's one of the things to decide.

Reply:
Added 25a475c with an example of functions changed using this way of SSL_CONNECTION_FROM_SSL model.

I suppose this is somewhat cleaner to outside reader of the code. One thing that concerned me was what if we wanted to add an ERR_raise into the error case. But that still can be done within the macro by using comma:

``` C
# define SSL_CONNECTION_FROM_SSL_ONLY(ssl) \
    (ssl->type == SSL_TYPE_SSL_CONNECTION  \
     ? (SSL_CONNECTION *)ssl               \
     : (ERR_raise(ERR_LIB_SSL, SSL_R_NOT_SSL_CONNECTION), NULL))
```